### DISCUSSION TITLE
Struggling with Setup Trying to use Ollama get Copilot API key is incorrect error

### DISCUSSION THREAD

**Discussion by jonstump** [olimorris/codecompanion.nvim#348] [Q&A] [answered]
I want to use my own ollama server and I've read through the adapters and even looked over various config examples, but I keep getting the error that I don't have a copilot api key, which makes sense as I don't have one entered, but I don't want to use copilot. I don't reference it in my setup for this.

I don't understand why this keeps giving me copilot api key errors.

```
return {
    "olimorris/codecompanion.nvim",
    config = function()
        require("codecompanion").setup({
            adapters = {
                ollama = function()
                    return require("codecompanion.adapters").extend("ollama", {
                        env = {
                            url = "https://localserverurl",
                            api_key = "APIKEY",
                        },
                        headers = {
                            ["Content-Type"] = "application/json",
                            ["Authorization"] = "Bearer ${api_key}",
                        },
                        parameters = {
                            sync = true,
                        },
                    })
                end,
            },
            strategies = {
                chat = {
                    adapter = "ollama",
                },
                inline = { adapter = "ollama" },
                agent = {
                    adapter = "ollama",
                    tools = {
                        opts = {
                            auto_submit_errors = true,
                        },
                    },
                },
            },
        })
    end,
}
```

**Comment by jonstump** [comment, marked as answer]
It appears I had the plugin referenced twice in my setup. This is why it was defaulting and then not hitting my setup. Once I removed the duplicate this started working.

**Reply by olimorris** [reply]
I also have a [minimal.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/minimal.lua) file in the repository if you need to troubleshoot.

**Reply by jonstump** [reply]
@olimorris I ended up using your personal dotfiles for reference which I found way way easier to understand than almost anything in the project on how to do the setup, including the file you just linked.  

I wish more projects like this would include links to personal dotfile repos because I usually end up having to track those down so I can see an actual real world working example of a setup vs example ones.

**Reply by olimorris** [reply]
Can you highlight the differences in the README to my dotfiles? They should be identical. Granted you don't get them in one full configuration block.

**Reply by jonstump** [reply]
The README just gives me a snippet of code, but doesn't give direction on where it should go. You just kind of present the adapters code, again without it in context of surrounding code. The adapters part in the readme seems to be "oh by the way'd" slapped on when it is one of the most crucial parts of setup. I was hard blocked on that for hours until I just tracked down your dotfiles.

Your example files in the project don't help me because they either are so huge it takes forever attempting to find what I might need or suffer the opposite fate, not detailed enough. _Your dotfiles represents a real world working setup. It's not example, it's a real thing._ There's a major different for my brain.

For me it is the difference between theory and reality. Your docs presents a theory of how it should work, your dotfiles presents a reality of it working. This is why I wish more people linked their working dotfiles.

**Reply by olimorris** [reply]
Were you unsure of where to place them in a Lazy.nvim setup?

A common example, providing a specific API key to an adapter, is noted in the README as:

```lua
require("codecompanion").setup({
  adapters = {
    openai = function()
      return require("codecompanion.adapters").extend("openai", {
        env = {
          api_key = "cmd:op read op://personal/OpenAI/credential --no-newline",
        },
      })
    end,
  },
})
```

and in my dotfiles as:

```lua
require("codecompanion").setup({
  adapters = {
    openai = function()
      return require("codecompanion.adapters").extend("openai", {
        env = {
          api_key = "cmd:op read op://personal/OpenAI_API/credential --no-newline",
        },
      })
    end,
  },
```

which is near identical.

**Reply by jonstump** [reply]
The other reason I wish more projects did this based on the simple argument that you're gonna keep your dotfiles up to date related to your project over your documentation as well. Documentation is a whole separate todo, where are your dotfiles **need** to work so you're incentivized to have them updated.

**Reply by jonstump** [reply]
I'll try adapting an analogy I've used in the past. Hopefully this helps as to why I'd like to see more people link their dotfiles in their READMEs as examples.

Imagine someone needs to learn how an engine runs because they're assembling a car.

README/Examples above - This is like showing someone just parts of an engine, not how they all fit together. If you've never seen an engine fully assembled and running it is probably gonna be a bad time only seeing a part and not realizing where it should go on a partially built engine. Or if you've specialized in NA engines and being shown a hybrid engine. Some of the parts might look familiar, but you may still be at a loss as to some of the electric parts.

Example files that are referenced - This is like showing a model engine. It doesn't work, but some of the parts move. While this might sound silly, it also isn't in reference to the rest of the car. Does the engine go in the front or the back? Is it a mid-engine? IDK I only have a model of a non-working engine and my car could be a front, mid, or rear engine. I know where some of those parts I have photos of go now, but where does it go in the car?

Your Dotfiles - This is akin to having access to a fully working car you can pop the hood on, turn over, take apart, etc. The entire puzzle is there put together, not lying in pieces. Seeing how it runs I could then place things appropriately and see where I had accidentally duplicated some code. Suddenly it would be like looking at a running car and realizing your car has two transmissions you didn't even know about because again, you haven't seen an actual running car and didn't know you had a spare transmission. You just kinda thought it needed two cause you had two.

**Reply by olimorris** [reply]
I get the analogy, I'm just trying to understand what parts of the README, specifically, could be improved. Hence why I pasted the comparison between my dotfiles and the README.

**Reply by jonstump** [reply]
Here are explicit things you could do to improve your readme.

1. Remove your note regarding the quick start guide because it isn't a quick start guide if it immediately has the user jumping around the readme out of order. It feels like it communicates "lulz we're kidding there is no quick start guide"
2. Replace that with the adapters setup. It's crucial to getting the tool running, why isn't it presented properly as such? You present it as **a note**, not the crucially needed step it is to setup. Its importance should be properly represented in this spot.
3. Include a link to your dotfiles as an example. Reverse Engineering a working thing works better for some of us vs jumping around looking for puzzle pieces.

I hope this helps.

**Reply by olimorris** [reply]
Appreciate your detailed feedback. Thank you.

**Comment by acascais** [comment]
@jonstump just out of curiosity, which ollama model do you run in your server and what are the server specs?

P.s: sry for the OT!

**Reply by jonstump** [reply]
IDK the full specs of the server, it's actually not mine but a roommate's. I do know that they are running Nvidia Tesla P40s and is some older Dell Enterprise server. It currently has a handful of models it can load, but I've largely stuck to ollama3.1:8b or 3.2.
